src/EGL/libeglvendor.c

```diff
@@ -0,0 +1,428 @@
+}
+
+void LoadVendorsFromConfigDir(const char *dirName)
+{
+    struct dirent **entries = NULL;
+    size_t dirnameLen;
+    const char *pathSep;
+    int count;
+    int i;
+
+    count = scandir(dirName, &entries, ScandirFilter, CompareFilenames);
+    if (count <= 0) {
+        return;
+    }
+
+    // Check if dirName ends with a "/" character. If it doesn't, then we need
+    // to add one when we construct the full file paths below.
+    dirnameLen = strlen(dirName);
+    if (dirName > 0 && dirName[dirnameLen - 1] != '/') {
+        pathSep = "/";
+    } else {
+        pathSep = "";
+    }
+
+    for (i=0; i<count; i++) {
+        char *path = NULL;
+        if (glvnd_asprintf(&path, "%s%s%s", dirName, pathSep, entries[i]->d_name) > 0) {
+            LoadVendorFromConfigFile(path);
+            free(path);
+        } else {
+            fprintf(stderr, "ERROR: Could not allocate vendor library path name\n");
+        }
+    }
+
+    free(entries);
+}
+
+void __eglInitVendors(void)
+{
+    glvnd_list_init(&__eglVendorList);
+}
+
+struct glvnd_list *__eglLoadVendors(void)
+{
+    __glvndPthreadFuncs.once(&loadVendorsOnceControl, LoadVendors);
+    return &__eglVendorList;
+}
+
+void __eglTeardownVendors(void)
+{
+    __EGLvendorInfo *vendor;
+    __EGLvendorInfo *vendorTemp;
+
+    glvnd_list_for_each_entry_safe(vendor, vendorTemp, &__eglVendorList, entry) {
+        glvnd_list_del(&vendor->entry);
+        TeardownVendor(vendor);
+    }
+}
+
+const __EGLapiExports __eglExportsTable = {
+    __eglThreadInitialize, // threadInit
+    __eglQueryAPI, // getCurrentApi
+    __eglGetCurrentVendor, // getCurrentVendor
+    __eglGetCurrentContext, // getCurrentContext
+    __eglGetCurrentDisplay, // getCurrentDisplay
+    __eglGetCurrentSurface, // getCurrentSurface
+    __eglFetchDispatchEntry, // fetchDispatchEntry
+    __eglSetError, // setEGLError
+    __eglSetLastVendor, // setLastVendor
+    __eglGetVendorFromDisplay, // getVendorFromDisplay
+    __eglGetVendorFromDevice, // getVendorFromDevice
+};
+
+void TeardownVendor(__EGLvendorInfo *vendor)
+{
+    if (vendor->glDispatch) {
+        __glDispatchDestroyTable(vendor->glDispatch);
+    }
+
+    /* Clean up the dynamic dispatch table */
+    if (vendor->dynDispatch != NULL) {
+        __glvndWinsysVendorDispatchDestroy(vendor->dynDispatch);
+        vendor->dynDispatch = NULL;
+    }
+
+    if (vendor->dlhandle != NULL) {
+        dlclose(vendor->dlhandle);
+    }
+
+    free(vendor);
+}
+
+static GLboolean LookupVendorEntrypoints(__EGLvendorInfo *vendor)
+{
+    memset(&vendor->staticDispatch, 0, sizeof(vendor->staticDispatch));
+
+    // TODO: A lot of these should be implemented (and probably generated) as
+    // normal EGL dispatch functions, instead of having to special-case them.
+
+#define LOADENTRYPOINT(ptr, name) do { \
+    *((__eglMustCastToProperFunctionPointerType *) &vendor->staticDispatch.ptr) = \
+        vendor->eglvc.getEGLProcAddress(name); \
```

Comment:
I can't get the type-punning warning to show up on my system. I'm using Fedora 24, with GCC 6.1.1 20160621.

Regardless, just returning (void *) seems like a reasonable workaround on any system we'll ever care about. Especially since if assigning a void* to a function pointer didn't work, then dlsym would break.